### PR TITLE
DatasourceVariable: Reloads query editors when data source type is updated

### LINKED ISSUE #32143: Variables: Changing the data source type for a data source variable requires the dashboard to refresh before working
Dashboard use to reproduce: Dashboard here: https://play.grafana.org/d/eYkq9GwMz/variables-changing-the-data-source-type-for-a-data-source-variable-requires-the-dashboard-to-refresh-before-working-32143?orgId=1

Steps to reproduce:
1. Create a data source variable named `ds` that uses TestData as data source type
2. Create a panel that uses the data source variable, should appear as `${ds}` in the data source picker if step 1 is done correctly
4. Go to Dashboard Settings => Variables => ds => and change data source type to InfluxDB, variable options should update [OK]
5. Go back to dashboard with ESC/back button
6. Notice how the panel data hasn't updated [Might be OK]
7. Edit the panel
8. Notice how the panel data has updated [OK]
9. Notice how the data source has changed [OK]

Alternative:
1. Create a data source variable named `ds` that uses TestData as data source type
2. Create a panel that uses the data source variable, should appear as `${ds}` in the data source picker if step 1 is done correctly
4. Edit the panel
5. Go to Dashboard Settings => Variables => ds => and change data source type to InfluxDB, variable options should update [OK]
6. Go back to edit panel with ESC/back button
7. Notice how the panel data hasn't updated [Might be OK]
8. Notice how the data source hasn't changed to InfluxDB [Not OK]
9. Click the refresh button
10. Notice how the panel data has updated [OK]
11. Notice how the data source hasn't changed to InfluxDB [Not OK]
12. Exit panel edit
13. Edit the panel
14. Notice how the data source has changed to InfluxDB [OK]

Works! But I kinda noticed that the dashboard needed to be refreshed before the datasource name updates. Is that the expected behavior? @hugohaggmark 

https://user-images.githubusercontent.com/29557702/111641514-856e3500-87fd-11eb-9ae1-8486587ff17b.mov

_Originally posted by @thisisobate in https://github.com/grafana/grafana/issues/32090#issuecomment-801968329_

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
This was a pretty hard issue to fix because it involved detecting changes to the mutable panel model on several levels of components. 
Not sure the changes and complexity in this PR are worth the fix and I would appreciate if anyone could come up with a cleaner solution.

**Which issue(s) this PR fixes**:
Fixes #32143

**Special notes for your reviewer**:

### CODE REVIEW

File: packages/grafana-ui/src/utils/useForceUpdate.ts [hunk 1]
```diff
@@ -2,6 +2,7 @@ import { useState } from 'react';
 
 /** @internal */
 export function useForceUpdate() {
-  const [value, setValue] = useState(0); // integer state
-  return () => setValue(value + 1); // update the state to force render
+  // @ts-ignore ignoring the return value and using the callback setter instead
+  const [_, setValue] = useState(0); // integer state
+  return () => setValue((v) => v + 1); // update the state to force render
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/utils/useForceUpdate.ts:7]: I stumbled upon this during the hunt: why query editors wouldn't reload. I think this should be a better approach, WDYT?

File: public/app/features/variables/datasource/adapter.ts [hunk 2]
```diff
@@ -25,7 +25,7 @@ export const createDataSourceVariableAdapter = (): VariableAdapter<DataSourceVar
       }
       return false;
     },
-    setValue: async (variable, option, emitChanges = false) => {
+    setValue: async (variable, option, emitChanges = true) => {
```
Comment (hugohaggmark) [inline public/app/features/variables/datasource/adapter.ts:28]: This change makes sure that all panels are refreshed when we change data source variables.

**Review by torkelo — COMMENTED** [review]
Given the complexity to solve this and that this feels like a very edge case scenario (changing data source type for data source variable while panel edit is open). 

And there is a workaround, just exist panel edit and enter it again. I think I prefer not fixing this.

**Comment by hugohaggmark** [comment]
> Given the complexity to solve this and that this feels like a very edge case scenario (changing data source type for data source variable while panel edit is open).
> 
> And there is a workaround, just exist panel edit and enter it again. I think I prefer not fixing this.

I agree, it doesn't feel worth the complexity.

**Pull request closed without merging**